storage/aptosdb/src/utils/iterators.rs

```diff
@@ -165,13 +192,61 @@ impl<'a> PrefixedStateValueIterator<'a> {
         }
         Ok(None)
     }
+
+    fn next_by_index(&mut self) -> Result<Option<(StateKey, StateValue)>> {
+        if !self.is_finished {
+            while let Some(((state_key, version), _)) =
+                self.index_iter.as_mut().unwrap().next().transpose()?
+            {
+                // In case the previous seek() ends on the same key with version 0.
+                if Some(&state_key) == self.prev_key.as_ref() {
+                    continue;
+                }
+                // Cursor is currently at the first available version of the state key.
+                // Check if the key_prefix is a valid prefix of the state_key we got from DB.
+                if !self.key_prefix.is_prefix(&state_key)? {
+                    // No more keys matching the key_prefix, we can return the result.
+                    self.is_finished = true;
+                    break;
+                }
+
+                if version > self.desired_version {
+                    self.index_iter
+                        .as_mut()
+                        .unwrap()
+                        .seek(&(state_key.clone(), self.desired_version))?;
+                    continue;
+                }
+
+                self.prev_key = Some(state_key.clone());
+                // Seek to the next key - this can be done by seeking to the current key with version 0
+                self.index_iter
+                    .as_mut()
+                    .unwrap()
+                    .seek(&(state_key.clone(), 0))?;
+
+                if let Some(Some(state_value)) =
+                    self.db
+                        .db_shard(state_key.get_shard_id())
+                        .get::<StateValueSchema>(&(state_key.clone(), version))?
+                {
+                    return Ok(Some((state_key, state_value)));
+                }
```

Comment:
it's a db corruption if the `get()` returns `None`, right? (I mean shall we error out?)

Reply:
I originally wrote in that way but feel it is an overkill?

Reply:
well, maybe just unwrap()? not too verbose and the logic is clear

Reply:
I'm not conformable to write `unwrap` here because the index and the state value are not written in the same db transaction.

Reply:
exactly the reason we shouldn't just return None in the case of two dbs mismatch?

Reply:
Sure, changed to return an error.